bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/scanner/FlatRowAdapter.java

```diff
@@ -0,0 +1,150 @@
+    public void startRow(ByteString rowKey) {
+      this.currentKey = rowKey;
+    }
+
+    @Override
+    public void startCell(String family, ByteString qualifier, long timestamp, List<String> labels,
+        long size) {
+      this.family = family;
+      this.qualifier = qualifier;
+      this.timestamp = timestamp;
+      this.labels = labels;
+      this.value = ByteString.EMPTY;
+    }
+
+    @Override
+    public void cellValue(ByteString value) {
+      this.value = this.value.concat(value);
+    }
+
+    /**
+     * Adds a Cell to {@link Cell}'s map which is ordered by family. cells received from
+     * {@link RowBuilder} has ordering as:
+     *   <ul>
+     *     <li>family names clustered, but not sorted</li>
+     *     <li>qualifiers in each family cluster is sorted lexicographically</li>
+     *     <li>then descending by timestamp</li>
+     *   </ul>
+     * The end result will be that {@link Cell} are ordered as:
+     *    <ul>
+     *      <li>lexicographical by family</li>
+     *      <li>then lexicographical by qualifier</li>
+     *      <li>then descending by timestamp</li>
+     *    </ul>
+     * A flattened version of the {@link Cell} map will be sorted correctly.
+     */
+    @Override
+    public void finishCell() {
+      if (!Objects.equal(this.family, this.previousFamily)) {
+        previousFamily = this.family;
+        familyCells = new ArrayList<>();
+        cells.put(this.family, this.familyCells);
+      }
+
+      FlatRow.Cell cell  = new FlatRow.Cell(this.family, this.qualifier, this.timestamp,
+                    this.value, this.labels);
+      this.familyCells.add(cell);
+      totalCellCount++;
+  }
+
+    /**
+     * This method flattens the {@link #cells} which has a map of Lists keyed by family name.
+     * The {@link #cells} TreeMap is sorted lexicographically, and each List is sorted by
+     * qualifier in lexicographically ascending order, and timestamp in descending order.
+     *
+     * @return an object of HBase {@link FlatRow}.
+     */
+    @Override
+    public FlatRow finishRow() {
+      ImmutableList.Builder<FlatRow.Cell> combined = ImmutableList.builderWithExpectedSize(totalCellCount);
+      for (List<FlatRow.Cell> familyCellList : cells.values()) {
+        combined.addAll(familyCellList);
+      }
+
+      return new FlatRow(this.currentKey, combined.build());
+    }
+
+    @Override
+    public void reset() {
+      this.currentKey = null;
+      this.cells = new TreeMap<>();
+      this.family = null;
+      this.qualifier = null;
+      this.labels = null;
+      this.timestamp = 0L;
+      this.value = null;
+      this.previousFamily = null;
+      this.totalCellCount = 0;
+    }
+
+    @Override
+    public FlatRow createScanMarkerRow(ByteString rowKey) {
+      return new FlatRow(rowKey, ImmutableList.<Cell>of());
+    }
+  }
+
+  @Override
+  public boolean isScanMarkerRow(FlatRow row) {
+    return row.getCells() == null || row.getCells().isEmpty();
```

Comment:
getCells() can never be null

Reply:
`FlatRow` can be created without `List<FlatRow.Cell>`, there is no null check in its [constructor](https://github.com/GoogleCloudPlatform/cloud-bigtable-client/blob/master/bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/scanner/FlatRow.java#L195). 

Would this class be only used by internal APIs? shall I remove this null check & add `@InternalApi` annotation?

Reply:
FlatRows should only be instantiated here or the old row merger. We haven’t been good about marking apis as internal, but I think the ctor for flatrow should be internal

Reply:
Removed the null check for `row.getCells()` as it is would never be null here.